skiko/src/iosMain/kotlin/org/jetbrains/skiko/SkikoUIView.kt

```diff
@@ -117,57 +122,360 @@ class SkikoUIView : UIView, UIKeyInputProtocol {
-        skiaLayer?.skikoView?.onTouchEvent(events.toTypedArray())
+        sendTouchEventToSkikoView(touches, SkikoTouchEventKind.ENDED)
     }
 
     override fun touchesMoved(touches: Set<*>, withEvent: UIEvent?) {
         super.touchesMoved(touches, withEvent)
-        val events: MutableList<SkikoTouchEvent> = mutableListOf()
-        for (touch in touches) {
-            val event = touch as UITouch
-            val (x, y) = event.locationInView(null).useContents { x to y }
-            val timestamp = (event.timestamp * 1_000).toLong()
-            events.add(
-                SkikoTouchEvent(x, y, SkikoTouchEventKind.MOVED, timestamp, event)
-            )
-        }
-        skiaLayer?.skikoView?.onTouchEvent(events.toTypedArray())
+        sendTouchEventToSkikoView(touches, SkikoTouchEventKind.MOVED)
     }
 
     override fun touchesCancelled(touches: Set<*>, withEvent: UIEvent?) {
         super.touchesCancelled(touches, withEvent)
-        val events: MutableList<SkikoTouchEvent> = mutableListOf()
-        for (touch in touches) {
-            val event = touch as UITouch
+        sendTouchEventToSkikoView(touches, SkikoTouchEventKind.CANCELLED)
+    }
+
+    private fun sendTouchEventToSkikoView(touches: Set<*>, kind: SkikoTouchEventKind) {
+        val events = touches.map {
+            val event = it as UITouch
             val (x, y) = event.locationInView(null).useContents { x to y }
             val timestamp = (event.timestamp * 1_000).toLong()
-            events.add(
-                SkikoTouchEvent(x, y, SkikoTouchEventKind.CANCELLED, timestamp, event)
-            )
+            SkikoTouchEvent(x, y, kind, timestamp, event)
+        }.toTypedArray()
+        skiaLayer?.skikoView?.onTouchEvent(events)
+    }
+
+    override fun inputDelegate(): UITextInputDelegateProtocol? {
+        return _inputDelegate
+    }
+
+    override fun setInputDelegate(inputDelegate: UITextInputDelegateProtocol?) {
+        _inputDelegate = inputDelegate
+    }
+
+    /**
+     * Returns the text in the specified range.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614527-text
+     * @param range A range of text in a document.
+     * @return A substring of a document that falls within the specified range.
+     */
+    override fun textInRange(range: UITextRange): String? {
+        return skiaLayer?.skikoView?.input?.textInRange(range.toSkikoTextRange())
+    }
+
+    /**
+     * Replaces the text in a document that is in the specified range.
+     * https://developer.apple.com/documentation/uikit/uitextinput/1614558-replace
+     * @param range A range of text in a document.
+     * @param text A string to replace the text in range.
+     */
+    override fun replaceRange(range: UITextRange, withText: String) {
+        skiaLayer?.skikoView?.input?.replaceRange(range.toSkikoTextRange(), withText)
+    }
+
+    override fun setSelectedTextRange(selectedTextRange: UITextRange?) {
+        selectedTextRange?.let {
```

Comment:
We should use this code:
```
skiaLayer?.skikoView?.input?.setSelectedTextRange(selectedTextRange?.toSkikoTextRange())
```
Otherwise, if iOS sends null, we won't do anything.

Reply:
Sure, thanks!